File: core/src/net/brenig/pixelescape/render/overlay/GamePausedOverlay.java
package net.brenig.pixelescape.render.overlay;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.math.Interpolation;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.ImageTextButton;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import net.brenig.pixelescape.lib.Reference;
import net.brenig.pixelescape.lib.Utils;
import net.brenig.pixelescape.render.ui.general.HorizontalSpacer;
import net.brenig.pixelescape.render.ui.general.PixelDialog;
import net.brenig.pixelescape.render.ui.general.VerticalSpacer;
import net.brenig.pixelescape.render.ui.ingame.ScoreWidget;
import net.brenig.pixelescape.screen.GameScreen;

/**
 * Displays when game gets paused
 */
public class GamePausedOverlay extends OverlayWithUi implements InputProcessor {

	private static final float ANIM_TIME_GAME_OVER = 0.6F;
	private static final float ANIM_TIME_PAUSED = 0.4F;
	private static final float TIME_TO_WAIT = 1.2F;

	private final boolean isGameOver;

	private final int highscore;

	private float animationProgress = 0;

	public GamePausedOverlay(final GameScreen screen, final boolean isGameOver) {
		super(screen);
		highscore = screen.game.getUserData().getHighScore(screen.getGameMode());
		this.isGameOver = isGameOver;

		screen.game.getFont().getData().setScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE);

		Table table = stage.createHeadUiLayoutTable();

		if (!isGameOver) {
			ImageTextButton btnResume = new ImageTextButton("Resume", screen.game.getSkin(), "resume");
			btnResume.getImageCell().padRight(6).padBottom(4);
			btnResume.addListener(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					resumeGame();
				}
			});
			table.add(btnResume);
		}

		table.add(new HorizontalSpacer());
		table.add(Utils.addFullScreenButtonToTable(Utils.addSoundAndMusicControllerToLayout(screen.game, Utils.createUIHeadLayout(screen.game))));

		if (!isGameOver) {
			table.add(new ScoreWidget(screen));
		}

		Table content = stage.createContentUiLayoutTable();
		content.add(new VerticalSpacer());
		content.row().expandX().center();

		TextButton btnMainMenu = new TextButton("Main Menu", screen.game.getSkin());
		btnMainMenu.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				gotoMainMenu();

			}
		});
		content.add(btnMainMenu).right().bottom().padBottom(40).padRight(10);

		TextButton btnRestartGame = new TextButton("Retry", screen.game.getSkin());
		btnRestartGame.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				restartGame();
			}
		});
		btnRestartGame.setVisible(false);
		content.add(btnRestartGame).left().bottom().padBottom(40).padLeft(10).width(btnMainMenu.getWidth());

		stage.getRootTable().layout();

		btnMainMenu.setVisible(false);
		btnRestartGame.setVisible(false);
		float menuX = btnMainMenu.getX();
		float restartX = btnRestartGame.getX();
		btnMainMenu.addAction(Actions.sequence(
				Actions.delay(isGameOver ? TIME_TO_WAIT : 0),
				Actions.moveTo(menuX + 1000, btnMainMenu.getY()),
				Actions.visible(true),
				Actions.moveTo(menuX, btnMainMenu.getY(), 0.8F, Interpolation.swing)));
		btnRestartGame.addAction(Actions.sequence(
				Actions.delay(isGameOver ? TIME_TO_WAIT + 0.2F : 0.2F),
				Actions.moveTo(restartX + 500, btnRestartGame.getY()),
				Actions.visible(true),
				Actions.moveTo(restartX, btnRestartGame.getY(), 0.8F, Interpolation.swing)));

	}

	@Override
	public void show() {
		screen.setOverlayInputProcessor(new InputMultiplexer(stage.getInputProcessor(), this));
		if (isGameOver) {
			screen.game.getGameMusic().fadeOutToStop(0.6F);
		} else {
			screen.game.getGameMusic().fadeOutToPause();
		}
	}


	@Override
	public void renderFirst(float delta) {
		//noinspection PointlessBooleanExpression,ConstantConditions
		if (Reference.SCREEN_TINT_STRENGTH > 0 && animationProgress > 0) {
			renderScreenTint(Utils.easeOut(animationProgress, isGameOver ? ANIM_TIME_GAME_OVER : ANIM_TIME_PAUSED, 2) * Reference.SCREEN_TINT_STRENGTH);
		}
	}

	@Override
	public void render(float delta) {
		//Game Paused
		screen.game.getRenderManager().begin();
		screen.game.getFont().getData().setScale(2, 4);
		if (isGameOver) {
			screen.game.getFont().setColor(1, 0, 0, 1);
			screen.getFontLayout().setText(screen.game.getFont(), "Game Over!");
		} else {
			screen.game.getFont().setColor(0, 0, 1, 1);
			screen.getFontLayout().setText(screen.game.getFont(), "Game Paused!");
		}

		//Slide in
		float gameOverAnim = isGameOver ? Math.max(0, screen.getWorld().getWorldHeight() / 2 - screen.getWorld().getWorldHeight() / 2 * Utils.easeOut(animationProgress, ANIM_TIME_GAME_OVER, 2)) : 0;
		float xPos = screen.getWorld().getWorldWidth() / 2 - screen.getFontLayout().width / 2;
		float txtGameOverHeight = screen.getFontLayout().height / 2;
		float yPos = ((2 * screen.getWorld().getWorldHeight()) / 3) + screen.getUiPos() + txtGameOverHeight + gameOverAnim;
		screen.game.getFont().draw(screen.game.getBatch(), screen.getFontLayout(), xPos, yPos);

		//Score
		screen.game.getFont().setColor(0, 1, 0, 1);
		screen.game.getFont().getData().setScale(1.2F);
		screen.getFontLayout().setText(screen.game.getFont(), "Your score: " + screen.getWorld().getPlayer().getScore());
		xPos = screen.getWorld().getWorldWidth() / 2 - screen.getFontLayout().width / 2;
		float txtScoreHeight = screen.getFontLayout().height / 2;
		yPos -= txtGameOverHeight + screen.game.getFont().getLineHeight() + txtScoreHeight;
		screen.game.getFont().draw(screen.game.getBatch(), screen.getFontLayout(), xPos, yPos);

		//Highscore
		if (isGameOver && highscore < screen.getWorld().getPlayer().getScore()) {
			screen.game.getFont().setColor(0, 1, 0, 1);
			screen.game.getFont().getData().setScale(1.2F);
			screen.getFontLayout().setText(screen.game.getFont(), "New Highscore!");
		} else {
			screen.game.getFont().setColor(0, 0, 1, 1);
			screen.game.getFont().getData().setScale(1.0F);
			screen.getFontLayout().setText(screen.game.getFont(), "Highscore: " + highscore);
		}
		xPos = screen.getWorld().getWorldWidth() / 2 - screen.getFontLayout().width / 2;
		float txtHighscoreHeight = screen.getFontLayout().height / 2;
		yPos -= screen.game.getFont().getLineHeight() + txtHighscoreHeight;
		screen.game.getFont().draw(screen.game.getBatch(), screen.getFontLayout(), xPos, yPos);

		animationProgress += delta;

		super.render(delta);
	}

	private void restartMusic() {
		if (isGameOver) {
			screen.game.getGameMusic().setCurrentMusic(screen.getGameMusic());
		}
		screen.game.getGameMusic().play(true);
	}

	@Override
	public boolean doesPauseGame() {
		return true;
	}

	@Override
	public boolean canHideCursor() {
		return false;
	}

	@Override
	public boolean keyDown(int keycode) {
		if (!isGameOver || animationProgress > TIME_TO_WAIT) {
			switch (keycode) {
				case Input.Keys.SPACE:
					resumeGame();
					return true;
				case Input.Keys.ESCAPE:
					gotoMainMenu();
					return true;

			}
		}
		return false;
	}

	@Override
	public boolean keyUp(int keycode) {
		return false;
	}

	@Override
	public boolean keyTyped(char character) {
		return false;
	}

	@Override
	public boolean touchDown(int screenX, int screenY, int pointer, int button) {
		return false;
	}

	@Override
	public boolean touchUp(int screenX, int screenY, int pointer, int button) {
		return false;
	}

	@Override
	public boolean touchDragged(int screenX, int screenY, int pointer) {
		return false;
	}

	@Override
	public boolean mouseMoved(int screenX, int screenY) {
		return false;
	}

	@Override
	public boolean scrolled(int amount) {
		return false;
	}

	private void resumeGame() {
		screen.setOverlay(new CountDownOverlay(screen));
		restartMusic();
	}

	/**
	 * display dialog asking to go to main menu
	 */
	private void gotoMainMenu() {
		if (isGameOver) {
			screen.showMainMenu();
		} else {
			final PixelDialog dialog = new PixelDialog("Sure?", screen.game.getSkin());
			dialog.setMovable(false);
			dialog.setModal(true);
			dialog.setPrefWidth(stage.getStageViewport().getWorldWidth() * 0.7F);
			dialog.setWidth(stage.getStageViewport().getWorldWidth() * 0.7F);
			dialog.label("Quit to main menu?");
			dialog.buttonYes(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					dialog.hide();
					screen.showMainMenu();
				}
			});
			dialog.buttonNo(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					dialog.hide();
				}
			});
			dialog.init();
			dialog.show(stage.getUiStage());
		}
	}

	private void restartGame() {
		if (isGameOver) {
			restartGameDo();
		} else {
			final PixelDialog dialog = new PixelDialog("Sure?", screen.game.getSkin());
			dialog.setMovable(false);
			dialog.setModal(true);
			dialog.setPrefWidth(stage.getStageViewport().getWorldWidth() * 0.7F);
			dialog.setWidth(stage.getStageViewport().getWorldWidth() * 0.7F);
			dialog.label("Restart game?");
			dialog.buttonYes(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					dialog.hide();
					restartGameDo();
				}
			});
			dialog.buttonNo(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					dialog.hide();
				}
			});
			dialog.init();
			dialog.show(stage.getUiStage());
		}
	}

	private void restartGameDo() {
		screen.resetToEmptyOverlay();
		screen.restart();
		restartMusic();
	}

	@Override
	protected boolean switchToPausedOverlayOnFocusChange() {
		return false;
	}
}

File: core/src/main/kotlin/net/brenig/pixelescape/render/overlay/GamePausedOverlay.kt
package net.brenig.pixelescape.render.overlay

import com.badlogic.gdx.Input
import com.badlogic.gdx.InputMultiplexer
import com.badlogic.gdx.InputProcessor
import com.badlogic.gdx.math.Interpolation
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.ui.ImageTextButton
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import net.brenig.pixelescape.game.data.constants.Reference
import net.brenig.pixelescape.game.data.constants.StyleNames
import net.brenig.pixelescape.lib.LangKeys
import net.brenig.pixelescape.lib.translate
import net.brenig.pixelescape.lib.utils.AnimationUtils
import net.brenig.pixelescape.lib.utils.UiUtils
import net.brenig.pixelescape.render.ui.general.HorizontalSpacer
import net.brenig.pixelescape.render.ui.general.PixelDialog
import net.brenig.pixelescape.render.ui.general.VerticalSpacer
import net.brenig.pixelescape.render.ui.ingame.ScoreWidget
import net.brenig.pixelescape.screen.GameScreen

/**
 * Displays when game gets paused
 */
class GamePausedOverlay(screen: GameScreen, private val isGameOver: Boolean) : OverlayWithUi(screen), InputProcessor {

    private val highscore: Int = screen.game.userData.getHighScore(screen.gameMode)

    private var animationProgress = 0f

    init {

        screen.game.font.data.setScale(1f)

        val table = stage.createHeadUiLayoutTable()

        if (!isGameOver) {
            val btnResume = ImageTextButton(LangKeys.Ingame.RESUME.translate(), screen.game.skin, StyleNames.BUTTON_RESUME)
            btnResume.imageCell.padRight(6f)
            btnResume.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    resumeGame()
                }
            })
            table.add(btnResume)
        }

        table.add(HorizontalSpacer())
        table.add(UiUtils.addFullScreenButtonToTable(UiUtils.addSoundAndMusicControllerToLayout(screen.game, UiUtils.createUIHeadLayout(screen.game))))

        if (!isGameOver) {
            table.add(ScoreWidget(screen))
        }

        val content = stage.createContentUiLayoutTable()
        content.add(VerticalSpacer())

        if (screen.game.gameConfig.gameService.isSessionActive) {
            content.row().expandX().center()

            val btnLeaderboard = UiUtils.createLeaderboardsButton(screen.game, stage, screen::gameMode)
            content.add(btnLeaderboard).padBottom(10f).expandX().center()
            btnLeaderboard.color.a = 0f
            btnLeaderboard.addAction(Actions.sequence(
                    Actions.scaleTo(0.05f, 0.05f),
                    Actions.moveBy(0F, -20F),
                    Actions.delay((if (isGameOver) TIME_TO_WAIT else 0F) + 0.8f),
                    Actions.parallel(
                            Actions.fadeIn(0.2f, Interpolation.fade),
                            Actions.scaleTo(1f, 1f, 0.2f, Interpolation.fade),
                            Actions.moveBy(0F, 20F, 0.2F))
            ))
        }

        content.row().expandX().center()

        val buttonPanel = Table()
        content.add(buttonPanel)

        val btnMainMenu = TextButton(LangKeys.Ingame.MAIN_MENU.translate(), screen.game.skin)
        btnMainMenu.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                gotoMainMenu()

            }
        })
        buttonPanel.add(btnMainMenu).right().bottom().padBottom(40f).padRight(10f)

        val btnRestartGame = TextButton(LangKeys.Ingame.RETRY.translate(), screen.game.skin)
        btnRestartGame.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                restartGame()
            }
        })
        btnRestartGame.isVisible = false
        buttonPanel.add(btnRestartGame).left().bottom().padBottom(40f).padLeft(10f).width(btnMainMenu.width)

        stage.rootTable.layout()

        btnMainMenu.isVisible = false
        btnRestartGame.isVisible = false
        val menuX = btnMainMenu.x
        val restartX = btnRestartGame.x
        btnMainMenu.addAction(Actions.sequence(
                Actions.delay(if (isGameOver) TIME_TO_WAIT else 0F),
                Actions.moveTo(menuX + 1000, btnMainMenu.y),
                Actions.visible(true),
                Actions.moveTo(menuX, btnMainMenu.y, 0.8f, Interpolation.swing)))
        btnRestartGame.addAction(Actions.sequence(
                Actions.delay(if (isGameOver) TIME_TO_WAIT + 0.2f else 0.2f),
                Actions.moveTo(restartX + 500, btnRestartGame.y),
                Actions.visible(true),
                Actions.moveTo(restartX, btnRestartGame.y, 0.8f, Interpolation.swing)))

    }

    override fun show() {
        screen.setOverlayInputProcessor(InputMultiplexer(stage.inputProcessor, this))
        if (isGameOver) {
            screen.game.gameMusic.fadeOutToStop(0.6f)
        } else {
            screen.game.gameMusic.fadeOutToPause()
        }
    }


    override fun renderFirst(delta: Float) {
        //noinspection PointlessBooleanExpression,ConstantConditions
        if (Reference.SCREEN_TINT_STRENGTH > 0 && animationProgress > 0) {
            renderScreenTint(AnimationUtils.easeOut(animationProgress, if (isGameOver) ANIM_TIME_GAME_OVER else ANIM_TIME_PAUSED, 2) * Reference.SCREEN_TINT_STRENGTH)
        }
    }

    override fun render(delta: Float) {
        //Game Paused
        screen.game.renderManager.begin()
        screen.game.bigFont.data.setScale(1.6f, 3f)
        if (isGameOver) {
            screen.game.bigFont.setColor(1f, 0f, 0f, 1f)
            screen.fontLayout.setText(screen.game.bigFont, LangKeys.Ingame.Overlay.GAME_OVER.translate())
        } else {
            screen.game.bigFont.setColor(0f, 0f, 1f, 1f)
            screen.fontLayout.setText(screen.game.bigFont, LangKeys.Ingame.Overlay.GAME_PAUSED.translate())
        }

        //Slide in
        val gameOverAnim = if (isGameOver) Math.max(0f, screen.world.worldHeight / 2 - screen.world.worldHeight / 2 * AnimationUtils.easeOut(animationProgress, ANIM_TIME_GAME_OVER, 2)) else 0F
        var xPos = screen.world.worldWidth / 2 - screen.fontLayout.width / 2
        val txtGameOverHeight = screen.fontLayout.height / 2
        var yPos = (2 * screen.world.worldHeight / 3).toFloat() + screen.uiPos.toFloat() + txtGameOverHeight + gameOverAnim
        screen.game.bigFont.draw(screen.game.batch, screen.fontLayout, xPos, yPos)

        //Score
        screen.game.bigFont.setColor(0f, 1f, 0f, 1f)
        screen.game.bigFont.data.setScale(1f)
        screen.fontLayout.setText(screen.game.bigFont, LangKeys.Ingame.Overlay.YOUR_SCORE.translate(screen.world.player.score))
        xPos = screen.world.worldWidth / 2 - screen.fontLayout.width / 2
        val txtScoreHeight = screen.fontLayout.height / 2
        yPos -= txtGameOverHeight + screen.game.bigFont.lineHeight + txtScoreHeight
        screen.game.bigFont.draw(screen.game.batch, screen.fontLayout, xPos, yPos)

        //Highscore
        if (isGameOver && highscore < screen.world.player.score) {
            screen.game.bigFont.setColor(0f, 1f, 0f, 1f)
            screen.game.bigFont.data.setScale(1f)
            screen.fontLayout.setText(screen.game.bigFont, LangKeys.Ingame.Overlay.NEW_HIGHSCORE.translate())
        } else {
            screen.game.bigFont.setColor(0f, 0f, 1f, 1f)
            screen.game.bigFont.data.setScale(0.8f)
            screen.fontLayout.setText(screen.game.bigFont, LangKeys.Ingame.Overlay.HIGHSCORE.translate(highscore))
        }
        xPos = screen.world.worldWidth / 2 - screen.fontLayout.width / 2
        val txtHighscoreHeight = screen.fontLayout.height / 2
        yPos -= screen.game.bigFont.lineHeight + txtHighscoreHeight
        screen.game.bigFont.draw(screen.game.batch, screen.fontLayout, xPos, yPos)

        animationProgress += delta

        super.render(delta)
        screen.game.bigFont.data.setScale(1f)
    }

    private fun restartMusic() {
        if (isGameOver) {
            if (screen.game.gameConfig.musicAvailable) {
                screen.game.gameMusic.currentMusic = screen.gameMusic
            }
        }
        screen.game.gameMusic.play(true)
    }

    override fun doesPauseGame(): Boolean {
        return true
    }

    override fun canHideCursor(): Boolean {
        return false
    }

    override fun keyDown(keycode: Int): Boolean {
        if (!isGameOver || animationProgress > TIME_TO_WAIT) {
            when (keycode) {
                Input.Keys.SPACE -> {
                    if (isGameOver) {
                        restartGameDo()
                    } else {
                        resumeGame()
                    }
                    return true
                }
                Input.Keys.ESCAPE -> {
                    gotoMainMenu()
                    return true
                }
            }
        }
        return false
    }

    override fun keyUp(keycode: Int): Boolean {
        return false
    }

    override fun keyTyped(character: Char): Boolean {
        return false
    }

    override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        return false
    }

    override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        return false
    }

    override fun touchCancelled(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        return false
    }

    override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
        return false
    }

    override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
        return false
    }

    override fun scrolled(amountX: Float, amountY: Float): Boolean {
        return false
    }

    private fun resumeGame() {
        screen.setOverlay(CountDownOverlay(screen))
        restartMusic()
    }

    /**
     * display dialog asking to go to main menu
     */
    private fun gotoMainMenu() {
        if (isGameOver) {
            screen.showMainMenu()
        } else {
            val dialog = PixelDialog(LangKeys.Ingame.Dialog.MAIN_MENU_TITLE.translate(), screen.game.skin)
            dialog.isMovable = false
            dialog.isModal = true
            dialog.prefWidth = stage.stageViewport.worldWidth * 0.7f
            dialog.width = stage.stageViewport.worldWidth * 0.7f
            dialog.label(LangKeys.Ingame.Dialog.MAIN_MENU_TEXT.translate())
            dialog.buttonYes(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    dialog.hide()
                    screen.showMainMenu()
                }
            })
            dialog.buttonNo(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    dialog.hide()
                }
            })
            dialog.init()
            dialog.show(stage.uiStage)
        }
    }

    private fun restartGame() {
        if (isGameOver) {
            restartGameDo()
        } else {
            val dialog = PixelDialog(LangKeys.Ingame.Dialog.RESTART_TITLE.translate(), screen.game.skin)
            dialog.isMovable = false
            dialog.isModal = true
            dialog.prefWidth = stage.stageViewport.worldWidth * 0.7f
            dialog.width = stage.stageViewport.worldWidth * 0.7f
            dialog.label(LangKeys.Ingame.Dialog.RESTART_TEXT.translate())
            dialog.buttonYes(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    dialog.hide()
                    restartGameDo()
                }
            })
            dialog.buttonNo(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    dialog.hide()
                }
            })
            dialog.init()
            dialog.show(stage.uiStage)
        }
    }

    private fun restartGameDo() {
        screen.resetToEmptyOverlay()
        screen.restart()
        restartMusic()
    }

    override fun switchToPausedOverlayOnFocusChange(): Boolean {
        return false
    }

    companion object {

        private val ANIM_TIME_GAME_OVER = 0.6f
        private val ANIM_TIME_PAUSED = 0.4f
        private val TIME_TO_WAIT = 1.2f
    }
}
